src/tests/tracing/eventpipe/allocationsampling/allocationsampling.cs

```diff
@@ -0,0 +1,255 @@
+            ret = RunAllocationSamplers();
+            if (ret != 100)
+                return ret;
+
+            return 100;
+        }
+
+        private static int RunAllocationSamplers()
+        {
+            //// trigger GC to avoid one during the next tests
+            //GC.Collect();
+
+            // run the same allocations test with no sampling, AllocationTick and AllocationSampled
+            Stopwatch clock = new Stopwatch();
+            clock.Start();
+            int allocatedSize = AllocateObjects(InstanceCount * 50);
+            clock.Stop();
+            var duration = clock.ElapsedMilliseconds;
+            Logger.logger.Log($"       #GCs: {GC.CollectionCount(0)}");
+            Logger.logger.Log($"No Sampling: {InstanceCount * 50} instances allocated for {allocatedSize} bytes in {duration} ms");
+            _noSamplingDuration = duration;
+            Logger.logger.Log("-");
+
+            //// trigger GC to avoid one during the next tests
+            //GC.Collect();
+
+            var ret = IpcTraceTest.RunAndValidateEventCounts(
+                new Dictionary<string, ExpectedEventCount>() { { "Microsoft-Windows-DotNETRuntime", -1 } },
+                _eventGeneratingActionForAllocationsWithAllocationTick,
+                // GCKeyword (0x1): 0b1
+                new List<EventPipeProvider>() { new EventPipeProvider("Microsoft-Windows-DotNETRuntime", EventLevel.Verbose, 0b1) },
+                1024, _DoesTraceContainAllocationTickEvents, enableRundownProvider: false);
+            if (ret != 100)
+                return ret;
+            Logger.logger.Log("-");
+
+            //// trigger GC to avoid one during the next tests
+            //GC.Collect();
+
+            ret = IpcTraceTest.RunAndValidateEventCounts(
+                new Dictionary<string, ExpectedEventCount>() { { "Microsoft-Windows-DotNETRuntime", -1 } },
+                _eventGeneratingActionForAllocationsWithAllocationSampled,
+                // AllocationSamplingKeyword (0x80000000000): 0b1000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000
+                new List<EventPipeProvider>() { new EventPipeProvider("Microsoft-Windows-DotNETRuntime", EventLevel.Informational, 0x80000000000) },
+                1024, _DoesTraceContainAllocationSampledEvents, enableRundownProvider: false);
+            if (ret != 100)
+                return ret;
+
+            return 100;
+        }
+
+        private static int AllocateObjects(int instanceCount)
+        {
+            int size = 0;
+            for (int i = 0; i < instanceCount; i++)
+            {
+                byte[] bytes = new byte[35];
+                size += bytes.Length;
+            }
+
+            return size;
+        }
+
+        // 2000 instances of 1KB byte arrays should trigger around 20 events with the default 1/100KB sampling rate
+        const int InstanceCount = 2000;
+        const int MinExpectedEvents = 15;
```

Comment:
The expected variance in the distribution of results is going to be way too high for this threshold to be reliable in a test. If my back of the envelope estimate is accurate we've got a ~20% chance to see less than 15 events for 2MB of allocations. If the test is trying to validate that some events get generated I'd use MinExpectedEvents=1. If the test is trying to validate the number of events is near the predicted number then we need enough margin of error that the test could be run billions of times and never fail. Something like 2M instances and expecting to see 18-22K events might be in the right ballpark.

If you are trying to validate for statistical accuracy I think there are at least two other sources of bias right now:
- A byte array with 1024 elements in it will be 1048 bytes large on x64
- Our sampling strategy never reports the same object more than once. If we truly sampled every byte with probability P = 1/100K then we'd expect some objects could be sampled multiple times. However rather than add that complexity to the runtime I think its easier to leave it out and let profilers weight the samples. For a given object of size N bytes that was reported once, the expected number of samples if we continued sampling bytes after the first hit is N*P/(1 - (1-P)^N). For these 1024 byte objects that is weight of about ~1.005.

The first bias causes the event count to be larger than expected by about 2%, the 2nd one causes the event count to be smaller than expected by about 0.5%. Neither one is too bad at array size 1024, but if we test with very small arrays the first error is giant and with very large arrays the 2nd error is giant.

Reply:
I started the test as simply validating that at least events were emitted with the right type name and size and then I wanted to approximate the number of events but, as you point out, this is flacky with such a small number of allocations.
I'm reverting to the original goal and I'll add new tests with more smaller objects.

Let me know if you have more tests ideas.